### PR TITLE
Extract trait interface for annotatable constructs and their builders

### PR DESCRIPTION
**Pull request by JakeWharton**
Part of #1553

### CODE REVIEW

File: kotlinpoet/src/main/java/com/squareup/kotlinpoet/FileSpec.kt [hunk 1]
```diff
@@ -248,7 +248,7 @@ public class FileSpec private constructor(
      * The annotation must either have a [`file` use-site target][AnnotationSpec.UseSiteTarget.FILE]
      * or not have a use-site target specified (in which case it will be changed to `file`).
      */
-    public fun addAnnotation(annotationSpec: AnnotationSpec): Builder = apply {
+    override fun addAnnotation(annotationSpec: AnnotationSpec): Builder = apply {
```
Comment (JakeWharton) [inline kotlinpoet/src/main/java/com/squareup/kotlinpoet/FileSpec.kt:251]: This override can be bypassed by mutating the collection directly. Should probably move the validation to the builder, but that's a separate issue.

File: kotlinpoet/api/kotlinpoet.api [hunk 2]
```diff
@@ -192,16 +205,14 @@ public final class com/squareup/kotlinpoet/FileSpec : com/squareup/kotlinpoet/Ta
...
 	public final fun addAliasedImport (Lcom/squareup/kotlinpoet/ClassName;Ljava/lang/String;Ljava/lang/String;)Lcom/squareup/kotlinpoet/FileSpec$Builder;
 	public final fun addAliasedImport (Lcom/squareup/kotlinpoet/MemberName;Ljava/lang/String;)Lcom/squareup/kotlinpoet/FileSpec$Builder;
 	public final fun addAliasedImport (Ljava/lang/Class;Ljava/lang/String;)Lcom/squareup/kotlinpoet/FileSpec$Builder;
 	public final fun addAliasedImport (Lkotlin/reflect/KClass;Ljava/lang/String;)Lcom/squareup/kotlinpoet/FileSpec$Builder;
-	public final fun addAnnotation (Lcom/squareup/kotlinpoet/AnnotationSpec;)Lcom/squareup/kotlinpoet/FileSpec$Builder;
-	public final fun addAnnotation (Lcom/squareup/kotlinpoet/ClassName;)Lcom/squareup/kotlinpoet/FileSpec$Builder;
-	public final fun addAnnotation (Ljava/lang/Class;)Lcom/squareup/kotlinpoet/FileSpec$Builder;
-	public final fun addAnnotation (Lkotlin/reflect/KClass;)Lcom/squareup/kotlinpoet/FileSpec$Builder;
+	public fun addAnnotation (Lcom/squareup/kotlinpoet/AnnotationSpec;)Lcom/squareup/kotlinpoet/FileSpec$Builder;
+	public synthetic fun addAnnotation (Lcom/squareup/kotlinpoet/AnnotationSpec;)Lcom/squareup/kotlinpoet/Taggable$Builder;
```
Comment (JakeWharton) [inline kotlinpoet/api/kotlinpoet.api:215]: Bug in _something_ here. Either the ABI tool or possibly the Kotlin compiler.

The return type of the synthetic should be `Annotatable$Builder`.

Comment (JakeWharton) [reply kotlinpoet/api/kotlinpoet.api:215]: Ah it was user error!

**Review by ZacSweers — APPROVED** [review]
Nice

**Review by Egorand — APPROVED** [review]

**Pull request merged**